tree/treeplayer/src/TTreeProcessorMT.cxx

```diff
@@ -517,51 +517,48 @@ void TTreeProcessorMT::Process(std::function<void(TTreeReader &)> func)
-   const auto &entries = clusterAndEntries.second;
-
-   // Retrieve number of entries for each file for each friend tree
-   const auto friendEntries = hasFriends ? GetFriendEntries(fFriendInfo) : std::vector<std::vector<Long64_t>>{};
-
-   // Parent task, spawns tasks that process each of the entry clusters for each input file
-   // TODO: for readability we should have two versions of this lambda, for shouldRetrieveAllClusters == true/false
-   auto processFile = [&](std::size_t fileIdx) {
-      // theseFiles contains either all files or just the single file to process
-      const auto &theseFiles = shouldRetrieveAllClusters ? fFileNames : std::vector<std::string>({fFileNames[fileIdx]});
-      // either all tree names or just the single tree to process
-      const auto &theseTrees = shouldRetrieveAllClusters ? fTreeNames : std::vector<std::string>({fTreeNames[fileIdx]});
-      // Evaluate clusters (with local entry numbers) and number of entries for this file, if needed
-      const auto theseClustersAndEntries =
-         shouldRetrieveAllClusters ? ClustersAndEntries{} : MakeClusters(theseTrees, theseFiles, maxTasksPerFile);
-
-      // All clusters for the file to process, either with global or local entry numbers
-      const auto &thisFileClusters = shouldRetrieveAllClusters ? clusters[fileIdx] : theseClustersAndEntries.first[0];
-
-      // Either all number of entries or just the ones for this file
-      const auto &theseEntries =
-         shouldRetrieveAllClusters ? entries : std::vector<Long64_t>({theseClustersAndEntries.second[0]});
-
+   // Per-file processing in case we retrieved all cluster info upfront
+   auto processFileUsingGlobalClusters = [&](std::size_t fileIdx) {
       auto processCluster = [&](const EntryCluster &c) {
-         auto r = fTreeView->GetTreeReader(c.start, c.end, theseTrees, theseFiles, fFriendInfo, fEntryList,
-                                           theseEntries, friendEntries);
+         auto r = fTreeView->GetTreeReader(c.start, c.end, fTreeNames, fFileNames, fFriendInfo, fEntryList, allEntries,
+                                           GetFriendEntries(fFriendInfo));
          func(*r);
       };
+      fPool.Foreach(processCluster, allClusters[fileIdx]);
+   };
 
-      fPool.Foreach(processCluster, thisFileClusters);
+   // Per-file processing that also retrieves cluster info for a file
+   auto processFileRetrievingClusters = [&](std::size_t fileIdx) {
+      // Evaluate clusters (with local entry numbers) and number of entries for this file
+      const auto &theseTreeNames = std::vector<std::string>({fTreeNames[fileIdx]});
+      const auto &theseFileNames = std::vector<std::string>({fFileNames[fileIdx]});
+      const auto theseClustersAndEntries = MakeClusters(theseTreeNames, theseFileNames, maxTasksPerFile);
+      const auto &theseClusters = theseClustersAndEntries.first[0];
+      const auto &theseEntries = theseClustersAndEntries.second[0];
```

Comment:
as we are not juggling both global and local things now, all the `"these"` are just noise:

```suggestion
      const auto &treeName = std::vector<std::string>({fTreeNames[fileIdx]});
      const auto &fileName = std::vector<std::string>({fFileNames[fileIdx]});
      const auto clustersAndEntries = MakeClusters(treeName, fileName, maxTasksPerFile);
      const auto &clusters = clustersAndEntries.first[0];
      const auto &entries = clustersAndEntries.second[0];
```

Reply:
Noted!

and is there a reason to have them local, but not global? To my understanding, it looks like `theseClusterAndEntries` could have been created before.

Reply:
the difference between the two lambdas is that in one case we retrieve all information for all trees upfront, and in the other (this case here) each task retrieves the information for the tree it cares about

Reply:
okay, then the next question is. why don't we do local retrieval always?

Reply:
to split that work between threads :)